### PR TITLE
Consolidate duplicate commands, establish HERMIT.API.Prelude

### PR DESCRIPTION
**Pull request by RyanGlScott**
This is my attempt at doing some slight cleanup of the command situation in `hermit-shell` at the moment. This pull request does the following:
- When there are two commands with the same name and number of arguments (but different argument types), it creates a typeclass to combine them. For example, `lhsT :: Transform LCore String -> Transform LCore String` and `lhsR :: Rewrite LCore -> Rewrite LCore` were combined into a single command `lhs :: LHSArgs a b => Transform a b -> Transform a b` (with `LHSArgs LCore String` and `LHSArgs LCore LCore` instances)
- Removes `OverloadedStrings` in certain parts of the API (namely, where it could confuse the newly introduced typeclasses). Things that were previously newtypes (`CoreString`, `Name`, `LemmaName`, etc.) are now type synonyms.
- Addresses the collision between certain `Prelude` functions and HERMIT commands. Both define the following: `abs`, `any`, `all`, `fail`, `id`, `log`, `not`, `repeat`, and `replicate`. I "fixed" this by introducing the `HERMIT.API.Prelude` module, which exports `HERMIT.API` and everything from `Prelude` except for those particular functions. If you want to use them, launching `hermit-shell` now runs `import qualified Prelude as P` by default.
- Changes all examples to use the above modifications

### CODE REVIEW

File: src/HERMIT/API/Dictionary/KURE.hs [hunk 1]
```diff
@@ -1,31 +1,89 @@
...
+--   Perform an identity rewrite.
+-- id :: Rewrite LCoreTC
+--   Perform an identity rewrite.
+id :: Id a => Rewrite a
+id = Transform $ method "idCore" []
+
+-- | Class of types that can have an identity rewrite via 'idCore'.
+class Id a
+instance Id LCore
+instance Id LCoreTC
```
Comment (xich) [inline src/HERMIT/API/Dictionary/KURE.hs]: This seems strange. Can we not just do away with this class (which has no methods) and make `id` fully polymorphic?

File: src/HERMIT/API/Dictionary/KURE.hs [hunk 2]
```diff
@@ -1,31 +1,89 @@
...
 
 -- | An always succeeding translation.
 success :: Transform LCore ()
 success = Transform $ method "success" []
 
 -- | A failing rewrite.
-fail_ :: String -> Rewrite LCore
-fail_ str = Transform $ method "fail_" [toJSON str]
+fail :: String -> Rewrite LCore
```
Comment (xich) [inline src/HERMIT/API/Dictionary/KURE.hs]: Shouldn't `fail` have type `String -> Rewrite a`?

File: src/HERMIT/API/Dictionary/Reasoning.hs [hunk 3]
```diff
@@ -173,29 +220,50 @@ extensionalityWith nm = Transform $ method "extensionalityWith" [toJSON nm]
...
+--   Apply a rewrite to the LHS of a quantified clause.
+lhs :: LHSArgs a b => Transform a b -> Transform a b
+lhs = lhsTransform
+
+class LHSArgs a b where
+    lhsTransform :: Transform a b -> Transform a b
+    lhsTransform t = Transform $ method "lhs" [toJSON t]
+
+instance LHSArgs LCore String
+instance LHSArgs LCore LCore
```
Comment (xich) [inline src/HERMIT/API/Dictionary/Reasoning.hs]: Why not:

```
lhs :: Transform a b -> Transform a b
lhs t = Transform $ method "lhs" [toJSON t]
```

(That is, why all the type classery if both instances use the same default method implementation anyway?)

**Comment by RyanGlScott** [comment]
My concern with making the above commands fully polymorphic was allowing them to be used in places where they shouldn't be. For example, AFAICT in HERMIT, `id` can only be used with a `RewriteT LCore` or `RewriteT LCoreTC`, so I wanted to avoid a scenario in `hermit-shell` where the phantom type of `Rewrite` could be instantiated with a type that `id` wasn't meant to work with.

This might be excessively paranoid, though, since in practice most of these commands can only be used with one specific `Shell` combinator. I'd have to investigate to see if there are any specific ways to circumvent HERMIT's types using `hermit-shell`.

**Comment by xich** [comment]
The fact that `id` only works at those two types is considered a limitation of the current shell... allowing more things (especially the KURE combinators like `id`, `any-td`, `>>>`, `<+`, etc) to be polymorphic was (at least IMHO) a goal for the new shell.

The reason we overloaded so much in the old shell (having two `lhs` commands, for instance) was to mitigate this limitation that everything had to be monomorphic. Having a single polymorphic `lhs` would be much better!

**Comment by RyanGlScott** [comment]
Oh, that's definitely a compelling argument, then. I'll go ahead and purge those typeclasses.

**Comment by RyanGlScott** [comment]
Closing stale PR.

**Pull request closed without merging**